omnigibson/utils/usd_utils.py

```diff
@@ -1659,6 +1660,104 @@ def mesh_prim_to_trimesh_mesh(mesh_prim, include_normals=True, include_texcoord=
+        edge_sizes (tuple of float): The sizes of the edges of the grid. The grid will have a checkerboard pattern
+            with the specified edge sizes. The largest edge size must be less than the extent.
+
+    Returns:
+        vertices (torch.Tensor): Vx3 vertices of the grid.
+        face_vert_indices (torch.Tensor): Fx3 indices of the vertices for each face of the grid.
+        face_uvs (torch.Tensor): Fx3x2 UV coordinates for each vertex in each face of the grid.
+    """
+    assert extent > max(edge_sizes), "Extent must be greater than the maximum edge size"
+    edge_sizes = sorted(list(edge_sizes) + [extent])
+
+    # Start a rolling list for vertices and face vertex indices.
+    vertices = []
+    face_vert_indices = []
+    face_uvs = []
+
+    # Iterate over cell sizes and the areas they will need to cover (which is the next cell size)
+    last_size_vert_idxes = None
+    for cell_size, area_size in zip(edge_sizes[:-1], edge_sizes[1:]):
+        # First, create the vertices for the area. These are named after the quadrants.
+        new_1 = (area_size, area_size, 0)
+        new_2 = (-area_size, area_size, 0)
+        new_3 = (-area_size, -area_size, 0)
+        new_4 = (area_size, -area_size, 0)
+        vertices.extend([new_1, new_2, new_3, new_4])
+
+        # Get the indices of the 3d coordinates for the area vertices
+        new_1_idx = len(vertices) - 4
+        new_2_idx = len(vertices) - 3
+        new_3_idx = len(vertices) - 2
+        new_4_idx = len(vertices) - 1
+
+        if last_size_vert_idxes is None:
+            # If this is the first iteration, we only have the area vertices. Connect them in two
+            # triangles to form the area.
+            face_vert_indices.extend([[new_1_idx, new_2_idx, new_3_idx], [new_1_idx, new_3_idx, new_4_idx]])
+
+            # For this case we can just get the indices of the UV coordinates for the area vertices.
+            repetitions = area_size / cell_size
+            new_1_uv = (repetitions, repetitions)
+            new_2_uv = (0, repetitions)
+            new_3_uv = (0, 0)
+            new_4_uv = (repetitions, 0)
+            face_uvs.extend([[new_1_uv, new_2_uv, new_3_uv], [new_1_uv, new_3_uv, new_4_uv]])
+        else:
+            # Otherwise, we have to connect the area to the previous area. We do this by connecting
+            # the vertices of the previous area to the vertices of the current area in triangles.
+            old_1_idx, old_2_idx, old_3_idx, old_4_idx = last_size_vert_idxes
+
+            area_face_verts = [
+                [old_1_idx, new_1_idx, new_2_idx],
+                [old_1_idx, new_2_idx, old_2_idx],
+                [old_2_idx, new_2_idx, new_3_idx],
+                [old_2_idx, new_3_idx, old_3_idx],
+                [old_3_idx, new_3_idx, new_4_idx],
+                [old_3_idx, new_4_idx, old_4_idx],
+                [old_4_idx, new_4_idx, new_1_idx],
+                [old_4_idx, new_1_idx, old_1_idx],
```

Comment:
holy cow

Reply:
These form the triangles that create a kind of ring mesh that has the previous square on the inside and the new square on the outside. It's a hardcoded list, no easy way to generate it otherwise. Looks like this. ![alt text](image.png)

Reply:
![image](https://github.com/user-attachments/assets/f0ce3167-6964-4465-afc4-73a7626b2cde)